javascript/computer_science/hash_map_data_structure.md

```diff
@@ -0,0 +1,199 @@
+
+You probably understand by this point why we must write a good hashing function which eliminates as many collisions as possible. Most likely you will not be writing your own hash functions, as most languages have it built in. But understanding how hash functions work is important.
+
+### Growth of a hash table
+
+Let's talk about the growth of our buckets. We don't have infinite memory, so we can't have an infinite amount of buckets. We need to start somewhere, but starting too big is also a waste of memory if we're only going to have a hash map with a single value in it. So to deal with this issue, we should start with a small array for our buckets. We'll use an array size `16`.
+
+<div class="lesson-note lesson-note--tip" markdown="1">
+  Most programming languages start with the default size of `16` because it's a power of 2, which help with some techniques for performance that require bit manipulation for indexes.
+</div>
+
+How are we going to insert into those buckets when our hash function generates big numbers like `20353924`? We make use of the modulo `%` operation `given any number modulo by 16 we will get a number between 0 and 15`.
+For example, if we are to find the bucket where the value `"Manon"` will land, then we do the following:
+![hashing using hash code and modular operation example](./hash_map_data_structure/imgs/01.png)
+
+If we keep adding nodes into our buckets then the buckets will start filling up, but what is more important is we know for a fact that if almost all buckets have items in them, then collisions are guaranteed, it is mathematically impossible not to.
+
+Remember we don't want collisions, in a perfect world each bucket will either have 0 or 1 Node only, so we grow our buckets to have more chance that our Nodes will spread and not stack up in the same buckets. To grow our buckets, we create a new buckets list that is double the size of the old buckets list, then we copy all nodes over to the new buckets.
+
+#### When do we know that it's time to grow our buckets size
+
+To deal with this, our hash map class need to keep track of two new fields, the `capacity` and the `load factor`.
+
+- The `capacity` is the total number of buckets we currently have. Keeping track of this will let us know if our map has reached a certain threshold aka `load factor`,
+
+- The `load factor` is a number that we can assign our hash map to at the start. It's the factor that will determine when is it a good time to grow our buckets, for example a load factor of `0.75` means our hash map will need to grow its buckets when the capacity reaches 75% full. Setting it too low will consume too much memory by having too many empty buckets, while setting it too high will allow our buckets to have many collisions before we grow them. Usually a good balance of `0.75 to 1` is used.
```

Comment:
IMO we can remove the phrase. "Usually a good ...". I don't think it serves any purpose, what do you think?
```suggestion
- The `load factor` is a number that we can assign our hash map to at the start. It's the factor that will determine when is it a good time to grow our buckets, for example a load factor of `0.75` means our hash map will need to grow its buckets when the capacity reaches 75% full. Setting it too low will consume too much memory by having too many empty buckets, while setting it too high will allow our buckets to have many collisions before we grow them.
```

Reply:
It can be reworded, but I feel letting people know that there is a conventional `sweet spot` of where a good balanced load factor can be. Rather than people just deciding a random number to be their load factor

Reply:
I'm curious about the 0.75-1 figure. I feel it's just easier to reword it as "Many hash map implementations use a default load factor of `0.75`"

Reply:
Why wouldn't we recommend the load factor to be between 0.75 and 1? I don't understand the reason we trying to avoid that. Saying many hash map implementation uses 0.75 is also not very accurate, as the number varies, but it's usually between 0.75 and 1.
Rust: 0.9
Java : 0.75
C#: 1

Do you have any reason to believe that this number is not optimal between space and performance complexities?

Reply:
I just thought giving a range of load factors might overwhelm the learners. But thanks for digging in and providing load factors in different languages. Let's put it in. My wording suggestion: "Hash map implementations across various languages use a load factor between 0.75 and 1" You can tweak the wording of course.

Reply:
I'm looking at load factors in different languages. Go has 6.5 if I'm reading it right https://github.com/golang/go/blob/fbfc203/src/runtime/hashmap.go#L67

Reply:
6.5 felt way too big and gonna have on average 6 collisions per bucket on average so something felt wrong and I did some digging. I think they use a load factor differently than our implementation they go with 6.5 out of 8 buckets so around 81% you can see the math here with 13 (double of load factor) divided by 16 (double of 8 buckets)
 check out this line here https://github.com/golang/go/blob/9b4b3e5acca2dabe107fa2c3ed963097d78a4562/src/runtime/map.go#L69C1-L69C1
 
 If that is true then in our implementation Go would have a load factor of 0.8

Reply:
Cool cool, I really can't understand code written in these repos lol Thank God for the comments 😆 

 This article (https://rcoh.me/posts/hash-map-analysis/) does say 6.5 for Go. Interestingly Ruby does load factor of 0.5 and uses an array instead of linked lists for buckets https://github.com/ruby/ruby/blob/master/st.c#L9

Reply:
I still find it really hard to believe a language would allow up to 6 pairs per bucket, that is too much performance hit to save memory, especially now days that memory is cheap. Unless their implementation is vastly different than a normal hash map.